File: python/tink/integration/gcpkms/_gcp_kms_aead_test.py
# Copyright 2024 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Tests for tink.python.tink.integration.gcp_kms_aead."""

from absl.testing import absltest
from absl.testing import parameterized
from google.api_core import exceptions as core_exceptions
from google.cloud import kms_v1

from tink import core
from tink.integration.gcpkms import _gcp_kms_client


GCP_KEY_NAME = 'projects/p1/locations/global/keyRings/kr1/cryptoKeys/ck1'
PLAINTEXT = b'plaintext'
CIPHERTEXT = b'ciphertext'
ASSOCIATED_DATA = b'associated_data'
STRING_63 = 'A' * 63
STRING_64 = 'A' * 64
FAKE_PROJECT_ID = '~@#$%^&*()_+|}{POI}?><213:"L{O}µ÷åß∑' * 5


class CustomException(core_exceptions.GoogleAPIError):
  pass


class GcpKmsAeadTest(parameterized.TestCase):

  def setUp(self):
    super().setUp()
    absltest.mock.patch.object(kms_v1, 'KeyManagementServiceClient').start()

  def tearDown(self):
    absltest.mock.patch.stopall()
    super().tearDown()

  def test_client_null(self):
    with self.assertRaises(core.TinkError):
      _gcp_kms_client._GcpKmsAead(None, GCP_KEY_NAME)

  @parameterized.parameters(
      '',
      None,
      'wrong/kms/key/format',
      'projects/p1/locations/global/keyRings/kr1/cryptoKeys',
      'projects/p1/locations/global/keyRings/kr1/cryptoKeys/ck1/',
      'projects/p1/locations/global@/keyRings/kr1/cryptoKeys/ck1',
      'projects/p1/locations/global/keyRings/kr1@/cryptoKeys/ck1',
      'projects/p1/locations/global/keyRings/kr1/cryptoKeys/ck1@',
      'projects/p1/locations/' + STRING_64 + '/keyRings/kr1/cryptoKeys/ck1',
      'projects/p1/locations/global/keyRings/' + STRING_64 + '/cryptoKeys/ck1',
      'projects/p1/locations/global/keyRings/kr1/cryptoKeys/' + STRING_64,
      'projects/p1/locations/global/keyRings/kr1/cryptoKeys/ck1/cryptoKeyVersions/1',
      'gcprojects://projects/p1/locations/global/keyRings/kr1/cryptoKeys/ck1',
  )
  def test_key_name_format_wrong(self, key_name):
    with self.assertRaises(core.TinkError):
      _gcp_kms_client._GcpKmsAead(kms_v1.KeyManagementServiceClient(), key_name)

  def test_encryption_fails(self):
    kms_v1.KeyManagementServiceClient().encrypt.side_effect = CustomException()
    gcp_aead = _gcp_kms_client._GcpKmsAead(
        kms_v1.KeyManagementServiceClient(), GCP_KEY_NAME
    )
    with self.assertRaises(core.TinkError):
      gcp_aead.encrypt(CIPHERTEXT, ASSOCIATED_DATA)

  def test_decryption_fails(self):
    kms_v1.KeyManagementServiceClient().decrypt.side_effect = CustomException()
    gcp_aead = _gcp_kms_client._GcpKmsAead(
        kms_v1.KeyManagementServiceClient(), GCP_KEY_NAME
    )
    with self.assertRaises(core.TinkError):
      gcp_aead.decrypt(PLAINTEXT, ASSOCIATED_DATA)

  @parameterized.parameters(
      GCP_KEY_NAME,
      'projects/p1@comp!/locations/global/keyRings/kr1/cryptoKeys/ck1',
      'projects/' + FAKE_PROJECT_ID + '/locations/l1/keyRings/k1/cryptoKeys/c1',
      'projects/p1/locations/' + STRING_63 + '/keyRings/kr1/cryptoKeys/ck1',
      'projects/p1/locations/global/keyRings/' + STRING_63 + '/cryptoKeys/ck1',
      'projects/p1/locations/global/keyRings/kr1/cryptoKeys/' + STRING_63,
  )
  def test_encryption_works(self, key_name):
    kms_v1.KeyManagementServiceClient().encrypt.return_value = (
        kms_v1.types.EncryptResponse(name=GCP_KEY_NAME, ciphertext=CIPHERTEXT)
    )
    gcp_aead = _gcp_kms_client._GcpKmsAead(
        kms_v1.KeyManagementServiceClient(), key_name
    )
    ciphertext = gcp_aead.encrypt(PLAINTEXT, ASSOCIATED_DATA)
    self.assertEqual(ciphertext, CIPHERTEXT)

  def test_decryption_works(self):
    kms_v1.KeyManagementServiceClient().decrypt.return_value = (
        kms_v1.types.DecryptResponse(plaintext=PLAINTEXT)
    )
    gcp_aead = _gcp_kms_client._GcpKmsAead(
        kms_v1.KeyManagementServiceClient(), GCP_KEY_NAME
    )
    plaintext = gcp_aead.decrypt(CIPHERTEXT, ASSOCIATED_DATA)
    self.assertEqual(plaintext, PLAINTEXT)


if __name__ == '__main__':
  absltest.main()

File: java_src/src/test/java/com/google/crypto/tink/integration/gcpkms/GcpKmsAeadTest.java
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

package com.google.crypto.tink.integration.gcpkms;

import static com.google.common.truth.Truth.assertThat;
import static java.nio.charset.StandardCharsets.UTF_8;
import static java.util.Arrays.asList;
import static org.junit.Assert.assertThrows;

import com.google.api.gax.core.NoCredentialsProvider;
import com.google.api.gax.grpc.GrpcTransportChannel;
import com.google.api.gax.rpc.FixedTransportChannelProvider;
import com.google.cloud.kms.v1.DecryptRequest;
import com.google.cloud.kms.v1.DecryptResponse;
import com.google.cloud.kms.v1.EncryptRequest;
import com.google.cloud.kms.v1.EncryptResponse;
import com.google.cloud.kms.v1.KeyManagementServiceClient;
import com.google.cloud.kms.v1.KeyManagementServiceGrpc.KeyManagementServiceImplBase;
import com.google.cloud.kms.v1.KeyManagementServiceSettings;
import com.google.crypto.tink.Aead;
import com.google.crypto.tink.KeysetHandle;
import com.google.crypto.tink.aead.AeadConfig;
import com.google.crypto.tink.aead.PredefinedAeadParameters;
import com.google.protobuf.ByteString;
import io.grpc.ManagedChannel;
import io.grpc.inprocess.InProcessChannelBuilder;
import io.grpc.inprocess.InProcessServerBuilder;
import io.grpc.stub.StreamObserver;
import io.grpc.testing.GrpcCleanupRule;
import java.security.GeneralSecurityException;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

@RunWith(JUnit4.class)
public final class GcpKmsAeadTest {

  // Implement a fake KeyManagementService.
  public static class FakeKmsImpl extends KeyManagementServiceImplBase {
    private static final Aead aead = createAead();

    private static final Aead createAead() {
      try {
        return KeysetHandle.generateNew(PredefinedAeadParameters.AES256_GCM)
            .getPrimitive(Aead.class);
      } catch (GeneralSecurityException e) {
        throw new RuntimeException(e);
      }
    }

    @Override
    public void encrypt(EncryptRequest request, StreamObserver<EncryptResponse> responseObserver) {
      EncryptResponse.Builder builder = EncryptResponse.newBuilder().setName(request.getName());
      try {
        byte[] ciphertext =
            aead.encrypt(
                request.getPlaintext().toByteArray(),
                request.getAdditionalAuthenticatedData().toByteArray());
        builder.setCiphertext(ByteString.copyFrom(ciphertext));
        responseObserver.onNext(builder.build());
        responseObserver.onCompleted();
      } catch (GeneralSecurityException e) {
        responseObserver.onError(e);
      }
    }

    @Override
    public void decrypt(DecryptRequest request, StreamObserver<DecryptResponse> responseObserver) {
      DecryptResponse.Builder builder = DecryptResponse.newBuilder();
      try {
        byte[] plaintext =
            aead.decrypt(
                request.getCiphertext().toByteArray(),
                request.getAdditionalAuthenticatedData().toByteArray());
        builder.setPlaintext(ByteString.copyFrom(plaintext));
        responseObserver.onNext(builder.build());
        responseObserver.onCompleted();
      } catch (GeneralSecurityException e) {
        responseObserver.onError(e);
      }
    }
  }

  private KeyManagementServiceClient kmsClient;

  /** This rule manages automatic graceful shutdown for the registered servers and channels. */
  @Rule public final GrpcCleanupRule grpcCleanup = new GrpcCleanupRule();

  @Before
  public void setUpClass() throws Exception {
    AeadConfig.register();

    // Create a server, add service, start, and register for automatic graceful shutdown.
    String serverName = InProcessServerBuilder.generateName();
    grpcCleanup.register(
        InProcessServerBuilder.forName(serverName)
            .directExecutor()
            .addService(new FakeKmsImpl())
            .build()
            .start());

    // Create a client channel and register for automatic graceful shutdown.
    ManagedChannel channel =
        grpcCleanup.register(InProcessChannelBuilder.forName(serverName).directExecutor().build());
    KeyManagementServiceSettings kmsSettings =
        KeyManagementServiceSettings.newBuilder()
            .setCredentialsProvider(NoCredentialsProvider.create())
            .setTransportChannelProvider(
                FixedTransportChannelProvider.create(
                    GrpcTransportChannel.newBuilder().setManagedChannel(channel).build()))
            .build();
    kmsClient = KeyManagementServiceClient.create(kmsSettings);
  }

  @Test
  public void kmsAead_works() throws Exception {
    String keyName = "projects/tink-test/locations/global/keyRings/unit-test/cryptoKeys/aead-key";
    FakeCloudKms fakeKms = new FakeCloudKms(asList(keyName));

    byte[] plaintext = "plaintext".getBytes(UTF_8);
    byte[] associatedData = "associatedData".getBytes(UTF_8);

    Aead kmsAead = new GcpKmsAead(fakeKms, keyName);

    byte[] ciphertext = kmsAead.encrypt(plaintext, associatedData);
    byte[] decrypted = kmsAead.decrypt(ciphertext, associatedData);
    assertThat(decrypted).isEqualTo(plaintext);

    byte[] associatedData2 = "associatedData2".getBytes(UTF_8);
    assertThrows(
        GeneralSecurityException.class, () -> kmsAead.decrypt(ciphertext, associatedData2));

    ciphertext[7] = (byte) (ciphertext[7] ^ 42);
    assertThrows(GeneralSecurityException.class, () -> kmsAead.decrypt(ciphertext, associatedData));
  }

  @Test
  public void builderWithSetCloudKms_sameAsPublicConstructor() throws Exception {
    String keyName = "projects/tink-test/locations/global/keyRings/unit-test/cryptoKeys/aead-key";
    FakeCloudKms fakeKms = new FakeCloudKms(asList(keyName));

    byte[] plaintext = "plaintext".getBytes(UTF_8);
    byte[] associatedData = "associatedData".getBytes(UTF_8);

    Aead kmsAead1 = new GcpKmsAead(fakeKms, keyName);
    Aead kmsAead2 = GcpKmsAead.builder().setCloudKms(fakeKms).setKeyName(keyName).build();

    byte[] ciphertext = kmsAead1.encrypt(plaintext, associatedData);
    byte[] decrypted = kmsAead2.decrypt(ciphertext, associatedData);
    assertThat(decrypted).isEqualTo(plaintext);
  }

  @Test
  public void kmsAeadGrpc_keyNameNotSet() throws Exception {
    assertThrows(
        GeneralSecurityException.class,
        () -> GcpKmsAead.builder().setKeyManagementServiceClient(kmsClient).build());
  }

  @Test
  public void kmsAeadGrpc_keyNameIsEmpty() throws Exception {
    assertThrows(
        GeneralSecurityException.class,
        () -> GcpKmsAead.builder().setKeyName("").setKeyManagementServiceClient(kmsClient).build());
  }

  @Test
  public void kmsAeadGrpc_keyNameInWrongFormat() throws Exception {
    String keyName = "projects/cloudkms-test/locations/global/keyRings/TinkKmsLib/cryptoKeys";
    assertThrows(
        GeneralSecurityException.class,
        () ->
            GcpKmsAead.builder()
                .setKeyName(keyName)
                .setKeyManagementServiceClient(kmsClient)
                .build());
  }

  @Test
  public void kmsAeadGrpc_noKmsClientIsGiven() throws Exception {
    String keyName = "projects/cloudkms-test/locations/global/keyRings/TinkKmsLib/cryptoKeys/key1";
    assertThrows(
        GeneralSecurityException.class, () -> GcpKmsAead.builder().setKeyName(keyName).build());
  }

  @Test
  public void kmsAeadGrpc_bothKmsCliensAreGiven() throws Exception {
    String keyName = "projects/cloudkms-test/locations/global/keyRings/TinkKmsLib/cryptoKeys/key1";
    FakeCloudKms fakeKms = new FakeCloudKms(asList(keyName));
    assertThrows(
        GeneralSecurityException.class,
        () ->
            GcpKmsAead.builder()
                .setKeyName(keyName)
                .setCloudKms(fakeKms)
                .setKeyManagementServiceClient(kmsClient)
                .build());
  }

  @Test
  public void kmsAeadGrpc_works() throws Exception {
    String keyName = "projects/cloudkms-test/locations/global/keyRings/TinkKmsLib/cryptoKeys/key1";
    Aead kmsAead =
        GcpKmsAead.builder().setKeyName(keyName).setKeyManagementServiceClient(kmsClient).build();

    byte[] plaintext = "plaintext".getBytes(UTF_8);
    byte[] associatedData = "associatedData".getBytes(UTF_8);
    byte[] associatedData2 = "associatedData2".getBytes(UTF_8);
    byte[] ciphertext = kmsAead.encrypt(plaintext, associatedData);
    byte[] decrypted = kmsAead.decrypt(ciphertext, associatedData);

    assertThat(decrypted).isEqualTo(plaintext);
    assertThrows(
        GeneralSecurityException.class, () -> kmsAead.decrypt(ciphertext, associatedData2));
    ciphertext[7] = (byte) (ciphertext[7] ^ 42);
    assertThrows(GeneralSecurityException.class, () -> kmsAead.decrypt(ciphertext, associatedData));
  }

  @Test
  public void kmsAeadGrpc_worksWithBytes() throws Exception {
    String keyName = "projects/cloudkms-test/locations/global/keyRings/TinkKmsLib/cryptoKeys/key1";
    Aead kmsAead =
        GcpKmsAead.builder().setKeyName(keyName).setKeyManagementServiceClient(kmsClient).build();

    byte[] plaintext = {(byte) 1, (byte) 2, (byte) 3, (byte) 4, (byte) 5, (byte) 6};
    byte[] associatedData = "associatedData".getBytes(UTF_8);
    byte[] associatedData2 = "associatedData2".getBytes(UTF_8);
    byte[] ciphertext = kmsAead.encrypt(plaintext, associatedData);
    byte[] decrypted = kmsAead.decrypt(ciphertext, associatedData);

    assertThat(decrypted).isEqualTo(plaintext);
    assertThrows(
        GeneralSecurityException.class, () -> kmsAead.decrypt(ciphertext, associatedData2));
    ciphertext[2] = (byte) (ciphertext[2] ^ 42);
    assertThrows(GeneralSecurityException.class, () -> kmsAead.decrypt(ciphertext, associatedData));
  }

  @Test
  public void kmsAeadGrpc_encryptDecryptEmptyString_success() throws Exception {
    String keyName = "projects/cloudkms-test/locations/global/keyRings/TinkKmsLib/cryptoKeys/key1";
    Aead kmsAead =
        GcpKmsAead.builder().setKeyName(keyName).setKeyManagementServiceClient(kmsClient).build();

    byte[] plaintext = "".getBytes(UTF_8);
    byte[] associatedData = "associatedData".getBytes(UTF_8);
    byte[] ciphertext = kmsAead.encrypt(plaintext, associatedData);
    byte[] decrypted = kmsAead.decrypt(ciphertext, associatedData);
    assertThat(decrypted).isEqualTo(plaintext);
  }

  @Test
  public void kmsAead_encryptDecryptEmptyString_success() throws Exception {
    String keyName = "projects/tink-test/locations/global/keyRings/unit-test/cryptoKeys/aead-key";
    FakeCloudKms fakeKms = new FakeCloudKms(asList(keyName));

    byte[] plaintext = "".getBytes(UTF_8);
    byte[] associatedData = "associatedData".getBytes(UTF_8);

    Aead kmsAead = new GcpKmsAead(fakeKms, keyName);

    byte[] ciphertext = kmsAead.encrypt(plaintext, associatedData);
    byte[] decrypted = kmsAead.decrypt(ciphertext, associatedData);
    assertThat(decrypted).isEqualTo(plaintext);
  }

  @Test
  public void twoKmsAeads_canOnlyDecryptTheirOwnCiphertext() throws Exception {
    String keyName = "projects/tink-test/locations/global/keyRings/unit-test/cryptoKeys/aead-key";
    String keyName2 = "projects/tink-test/locations/global/keyRings/unit-test/cryptoKeys/aead-key2";
    FakeCloudKms fakeKms = new FakeCloudKms(asList(keyName, keyName2));

    byte[] plaintext = "plaintext".getBytes(UTF_8);
    byte[] associatedData = "associatedData".getBytes(UTF_8);

    Aead kmsAead = new GcpKmsAead(fakeKms, keyName);
    byte[] ciphertext = kmsAead.encrypt(plaintext, associatedData);

    Aead kmsAead2 = new GcpKmsAead(fakeKms, keyName2);
    byte[] ciphertext2 = kmsAead2.encrypt(plaintext, associatedData);

    assertThat(kmsAead.decrypt(ciphertext, associatedData)).isEqualTo(plaintext);
    assertThat(kmsAead2.decrypt(ciphertext2, associatedData)).isEqualTo(plaintext);

    assertThrows(
        GeneralSecurityException.class, () -> kmsAead2.decrypt(ciphertext, associatedData));
    assertThrows(
        GeneralSecurityException.class, () -> kmsAead.decrypt(ciphertext2, associatedData));
  }
}
